pkg/segment/set.go

```diff
@@ -932,11 +1001,203 @@ func (n *node) rebalanceAfterRemove(gap GapIterator) GapIterator {
 		}
 		p.children[p.nrSegments] = nil
 		p.nrSegments--
+		// Update maxGap of left locally, no need to change p and right because
+		// p's contents is not changed and right is already invalid.
+		left.updateLocalMaxGap()
 		// This process robs p of one segment, so recurse into rebalancing p.
 		n = p
 	}
 }
 
+// updateMaxGapLeaf updates maxGap bottom-up from the calling leaf until no necessary update.
+//
+// Precondition: Should always and only be called on a leaf node.
+func (n *node) updateMaxGapLeaf() {
+	var max Key
+	if !n.hasChildren {
+		for i := 0; i <= n.nrSegments; i++ {
+			currentGap := GapIterator{n, i}
+			if temp := currentGap.Range().Length(); i == 0 || temp > max {
+				max = temp
+			}
+		}
+	} else {
+		panic(fmt.Sprintf("updateMaxGapLeaf should always be called on leaf node: %v", n))
+	}
+	if max == n.maxGap {
+		// If new max equals the old maxGap, no update is needed.
+		return
+	} else if max > n.maxGap {
+		// If new max is larger than the old maxGap, update it.
+		n.maxGap = max
+		// Only if new maxGap is larger than parent's old maxGap, propagate this update to parent.
+		if n.parent != nil && n.parent.maxGap < n.maxGap {
+			n.parent.updateMaxGapInternal(max)
+		}
+	} else {
+		if n.parent != nil && n.parent.maxGap == n.maxGap {
+			// If new max is smaller than the old maxGap, and this gap used to be the maxGap of its parent,
+			// iterate parent's children and calculate parent's new maxGap.
+			// (It's probable that parent has two children with the old maxGap, but we need to check it anyway.)
+			n.maxGap = max
+			var parentNewMax Key
+			for i := 0; i <= n.parent.nrSegments; i++ {
+				child := n.parent.children[i]
+				if temp := child.maxGap; i == 0 || temp > parentNewMax {
+					parentNewMax = temp
+				}
+			}
+			// If parent's new maxGap differs from the old one, propagate this update.
+			if parentNewMax != n.parent.maxGap {
+				n.parent.updateMaxGapInternal(parentNewMax)
+			}
+		} else {
+			// If new max is smaller than the old maxGap, but parent's maxGap is not this one,
+			// only need to update locally
+			n.maxGap = max
+		}
+	}
+}
+
+// updateMaxGapInternal updates the maxGap on internal nodes recursively.
+//
+// Precondition: Should only be called by updateMaxGapLeaf or updateMaxGapInternal recursively.
+func (n *node) updateMaxGapInternal(newMaxGap Key) {
+	if !n.hasChildren {
+		panic(fmt.Sprintf("updateMaxGapInternal should always be called on non-leaf node: %v", n))
+	}
+	if newMaxGap == n.maxGap {
+		// If newMaxGap equals the old maxGap, no update is needed.
+		return
+	} else if newMaxGap > n.maxGap {
```

Comment:
See comments above, re: nesting. This should just be:

if newMaxGap == n.maxGap {
  ...
   return
}
if newMaxGap > n.maxGap {
   ..
}